### PR TITLE
added quick_sort_3

### PR DESCRIPTION
**Pull request by immortal-j**
adding quicksort3 as an faster alternate of quicksort

### CODE REVIEW

File: sorting/quick_sort_3.cpp [hunk 1]
```diff
@@ -0,0 +1,84 @@
+/*
+Author: immortal-j
+Implementation Details
+Quick sort 3 works on Dutch National Flag Algorithm
+The major diffrence between simple quicksort and quick sort 3 comes in the
+function partition3 In quick_sort_partition3 we divide the vector/array into 3
+parts. quick sort 3 works faster in some cases as compared to simple quicksort.
+*/
+#include <iostream>
+// function to swap elements
```
Comment (Panquesito7) [inline sorting/quick_sort_3.cpp:10]: Please use a comment structure like:

```c++
/**
 * Something here
 * Something else here
 */
```

File: sorting/quick_sort_3.cpp [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+void swap(int *a, int *b) {
+  int temp = *a;
+  *a = *b;
+  *b = temp;
+}
+
+// function to print an array
```
Comment (Panquesito7) [inline sorting/quick_sort_3.cpp:17]: Same here.

File: sorting/quick_sort_3.cpp [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+/* This function partitions arr[] in three parts
+a) arr[l..i] contains all elements smaller than pivot
+b) arr[i+1..j-1] contains all occurrences of pivot
+c) arr[j..r] contains all elements greater than pivot */
+void partition3(int arr[], int low, int high, int *i, int *j) {
+  // To handle 2 elements
+  if (high - low <= 1) {
+    if (arr[high] < arr[low])
```
Comment (Panquesito7) [inline sorting/quick_sort_3.cpp:31]:
```suggestion
    if (arr[high] < arr[low]) {
```

File: sorting/quick_sort_3.cpp [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+      swap(&arr[high], &arr[low]);
+    *i = low;
```
Comment (Panquesito7) [inline sorting/quick_sort_3.cpp:33]:
```suggestion
    }
    *i = low;
```

File: sorting/quick_sort_3.cpp [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+  int mid = low;
+  int pivot = arr[high];
+  while (mid <= high) {
+    if (arr[mid] < pivot)
+      swap(&arr[low++], &arr[mid++]);
+    else if (arr[mid] == pivot)
+      mid++;
+    else if (arr[mid] > pivot)
+      swap(&arr[mid], &arr[high--]);
```
Comment (Panquesito7) [inline sorting/quick_sort_3.cpp:46]:
```suggestion
      swap(&arr[mid], &arr[high--]);
    }
```

File: sorting/quick_sort_3.cpp [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  }
+
+  int mid = low;
+  int pivot = arr[high];
+  while (mid <= high) {
+    if (arr[mid] < pivot)
+      swap(&arr[low++], &arr[mid++]);
+    else if (arr[mid] == pivot)
+      mid++;
+    else if (arr[mid] > pivot)
```
Comment (Panquesito7) [inline sorting/quick_sort_3.cpp:45]:
```suggestion
    } else if (arr[mid] > pivot) {
```

File: sorting/quick_sort_3.cpp [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    *j = high;
+    return;
+  }
+
+  int mid = low;
+  int pivot = arr[high];
+  while (mid <= high) {
+    if (arr[mid] < pivot)
+      swap(&arr[low++], &arr[mid++]);
+    else if (arr[mid] == pivot)
```
Comment (Panquesito7) [inline sorting/quick_sort_3.cpp:43]:
```suggestion
    } else if (arr[mid] == pivot) {
```

File: sorting/quick_sort_3.cpp [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+      swap(&arr[high], &arr[low]);
+    *i = low;
+    *j = high;
+    return;
+  }
+
+  int mid = low;
+  int pivot = arr[high];
+  while (mid <= high) {
+    if (arr[mid] < pivot)
```
Comment (Panquesito7) [inline sorting/quick_sort_3.cpp:41]:
```suggestion
    if (arr[mid] < pivot) {
```

File: sorting/quick_sort_3.cpp [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+  int temp = *a;
+  *a = *b;
+  *b = temp;
+}
+
+// function to print an array
+void print_arr(int arr[], int n) {
+  for (int i = 0; i < n; ++i)
+    std::cout << arr[i] << " ";
+  std::cout << "\n";
```
Comment (Panquesito7) [inline sorting/quick_sort_3.cpp:21]:
```suggestion
  }
  std::cout << "\n";
```

File: sorting/quick_sort_3.cpp [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+// function to swap elements
+void swap(int *a, int *b) {
+  int temp = *a;
+  *a = *b;
+  *b = temp;
+}
+
+// function to print an array
+void print_arr(int arr[], int n) {
+  for (int i = 0; i < n; ++i)
```
Comment (Panquesito7) [inline sorting/quick_sort_3.cpp:19]:
```suggestion
  for (int i = 0; i < n; ++i) {
```

**Comment by kvedala** [comment]
This pull-request seems to be abandoned. I have taken the contribution and updated it to the specs and standards fo this repo and created #958 and closing this PR.

**Pull request closed without merging**